srml/staking/src/lib.rs

```diff
@@ -640,57 +696,141 @@ impl<T: Trait> Module<T> {
-		// cpu=O(V.log(s)) average, O(V.s) worst.
-		let count = Self::validator_count() as usize;
-		let candidates = if candidates.len() <= count {
-			candidates.sort_unstable_by(|a, b| b.1.total.cmp(&a.1.total));
-			candidates
-		} else {
-			candidates.into_iter().fold(vec![], |mut winners: Vec<(T::AccountId, Exposure<T::AccountId, BalanceOf<T>>)>, entry| {
-				if let Err(insert_point) = winners.binary_search_by_key(&entry.1.total, |i| i.1.total) {
-					if winners.len() < count {
-						winners.insert(insert_point, entry)
-					} else {
-						if insert_point > 0 {
-							// Big enough to be considered: insert at beginning and swap up to relevant point.
-							winners[0] = entry;
-							for i in 0..(insert_point - 1) {
-								winners.swap(i, i + 1)
-							}
+			Nominations {
+				who,
+				nominees: nominees.into_iter()
+					.map(|n| Vote {who: n, load: Perquill::zero(), backing_stake: BalanceOf::<T>::zero()})
+					.collect::<Vec<Vote<T::AccountId, BalanceOf<T>>>>(),
+				stake: nominator_stake,
+				load : Perquill::zero(),
+			}
+		}).collect::<Vec<Nominations<T::AccountId, BalanceOf<T>>>>();
+		
+		// 3- optimization: 
+		// candidates who have 0 stake => have no votes or all null-votes. best to kick them out not.
+		let mut candidates = candidates.into_iter().filter(|c| c.approval_stake > BalanceOf::<T>::zero())
+			.collect::<Vec<Candidate<T::AccountId, BalanceOf<T>>>>();
+
+		// 4- If we have more candidates then needed, run phragmen.
+		if candidates.len() > rounds {
+			// Main election loop
+			for _round in 0..rounds {
+				// Loop 1: initialize score
+				for nominaotion in &nominations {
+					for vote in &nominaotion.nominees {
+						let candidate = &vote.who;
+						if let Some(index) = candidates.iter().position(|i| i.who == *candidate) {
+							let approval_stake = candidates[index].approval_stake;
+							candidates[index].score = Perquill::from_xth(approval_stake.as_());
+						}
+					}
+				}
+				// Loop 2: increment score.
+				for nominaotion in &nominations {
+					for vote in &nominaotion.nominees {
+						let candidate = &vote.who;
+						if let Some(index) = candidates.iter().position(|i| i.who == *candidate) {
+							let approval_stake = candidates[index].approval_stake;
+							let temp =
+								nominaotion.stake.as_()
+								* nominaotion.load.extract()
+								/ approval_stake.as_();
+							candidates[index].score = Perquill::from_quilltionths(candidates[index].score.extract() + temp);
+						}
+					}
+				}
+
+				// Find the best
+				let (winner_index, _) = candidates.iter().enumerate().min_by_key(|&(_i, c)| c.score.extract())
+					.expect("candidates length is checked to be >0; qed");
+
+				// loop 3: update nominator and vote load
+				let winner = candidates.remove(winner_index);
+				for nominator_idx in 0..nominations.len() {
+					for vote_idx in 0..nominations[nominator_idx].nominees.len() {
+						if nominations[nominator_idx].nominees[vote_idx].who == winner.who {
```

Comment:
Don't use raw array access all the time, if not required.

Reply:
This code block is screaming "use iterators"